src/vm/analysis/type_checker/mod.rs

```diff
@@ -168,6 +168,27 @@ impl FunctionType {
             },
         }
     }
+
+    pub fn check_args_by_allowing_trait_cast(&self, func_args: &[Value]) -> CheckResult<TypeSignature> {
+        let (expected_args, returns) = match self {
+            FunctionType::Fixed(FixedFunction{ args, returns }) => (args, returns),
+            _ => panic!("Unexpected function type")
+        };
+        check_argument_count(expected_args.len(), func_args)?;
+
+        for (expected_arg, arg) in expected_args.iter().zip(func_args.iter()).into_iter() {
+            match (&expected_arg.signature, arg) {
+                (TypeSignature::TraitReferenceType(_), Value::Principal(PrincipalData::Contract(_))) => continue,
```

Comment:
Can this check the trait compliance? We should be able to pull the `ContractAnalysis` struct out of the VM database, which can be used to check trait compliance.

Reply:
I'm not sure to understand what you have in mind but if I do, I think the proposed alternative would not cover the "trait conformance by accident", would it?

Reply:
checking the trait compliance was part of my initial thinking, but I later changed my mind and managed to convince myself that something else would do it later.  I just checked today, and that's not the case.  If the contract you pass does not implement the trait, execution will fail, repeatedly as miners will keep trying to include the transaction.

Will look into adding a check.

Reply:
addressed in 9738843

Reply:
I think that the problem with this is approach, is that we're loosing the "trait compliance by. accident", which was a desired feature.
A contract that implements a method with a compatible signature must be able to be passed in as a trait <...> argument, without complying to the whole trait.
Taking your test, if you add a second method in your trait, that does not get its implementation in the contract being passed as a trait, then I think that the mempool will reject instead of granting access.

Reply:
I believe `.check_trait_compliance` will allow "trait compliance by accident" -- it makes sure that the provided contract defines the correct functions from the trait (whether or not the contract explicitly calls `impl-trait`). It's true that the contract needs to comply with the _whole_ trait, but I think that's a feature we want. Contracts which only implement a subset of the trait functions should not be allowed.

Reply:
agree with @kantai, the contract should implement the whole trait.  The current implementation does not check which method is being called by the called contract, not I don't think it should.  So `check_trait_compliance` provides exactly what's need to validate entry.
The latest version of `swapr` relies on that assumption, as it records the principal of the trait being passed at creation, and that's the only trait you can pass to that pair for the other functions.  If the trait was partially implemented, this would break.  In fact, if a contract requires a subset of a known trait, it could define its own sub-trait (if that's a word) and get that checked.